src/Mobs/Path.cpp

```diff
@@ -0,0 +1,391 @@
+		}
+		
+		// Calculate H. This is A*'s Heuristics value.
+		#if DISTANCE_MANHATTEN == 1
+		// Manhatten distance. DeltaX + DeltaY + DeltaZ.
+		a_Cell->m_H = 10 * (abs(a_Cell->m_Location.x-m_Destination.x) + abs(a_Cell->m_Location.y-m_Destination.y) + abs(a_Cell->m_Location.z-m_Destination.z));
+		#else
+		// Euclidian distance. sqrt(DeltaX^2 + DeltaY^2 + DeltaZ^2), more precise.
+		a_Cell->m_H = std::sqrt( (a_Cell->m_Location.x-m_Destination.x) * (a_Cell->m_Location.x-m_Destination.x) * 100+ (a_Cell->m_Location.y-m_Destination.y) * (a_Cell->m_Location.y-m_Destination.y) * 100 + (a_Cell->m_Location.z-m_Destination.z) * (a_Cell->m_Location.z-m_Destination.z) * 100);
+		#endif
+		
+		
+		#if HEURISTICS_ONLY == 1
+		a_Cell->m_F = a_Cell->m_H;  // Depth-first search(Might be the wrong name). Faster, can yeild paths that are far from optimal.
+		#else
+		a_Cell->m_F = a_Cell->m_H + a_Cell->m_G;  // Regular A*.
+		#endif
+		
+		OpenListAdd(a_Cell);
+		return;
+	}
+	
+
+	
+	// Case 3: Cell is in the open list, check if G and H need an update.
+	int NewG = a_Caller->m_G + a_GDelta;
+	if (NewG < a_Cell->m_G)
+	{
+		a_Cell->m_G = NewG;
+		a_Cell->m_H = a_Cell->m_F + a_Cell->m_G;
+		a_Cell->m_Parent = a_Caller;
+	}
+	
+}
+
+
+
+
+
+ePathFinderStatus cPath::Step()
+{
+	// printf("cPath::step() - Stepping...\n");
+	if (m_Status == ePathFinderStatus::CALCULATING)
+	{
+		if (m_StepsLeft == 0)
+		{
+			// printf("cPath::step() - No more steps left. Path either too far or non existent.\nIf the former, increase MaxSteps in constructor.\n");
+			FinishCalculation(ePathFinderStatus::PATH_NOT_FOUND);
+		}
+		else
+		{
+			--m_StepsLeft;
+			int i;
+			for (i = 0; i < CALCULATIONS_PER_STEP; ++i)
+			{
+				if (Step_Internal())  // Step_Internal returns true when no more calculation is needed.
+				{
+					break;  // if we're here, m_Status must have changed either to PATH_FOUND or PATH_NOT_FOUND.
+				}
+			}
+		}
+	}
+	return m_Status;
+}
+
+
+
+void cPath::ProcessIfWalkable(const Vector3d & a_Location, cPathCell * a_Parent, int a_Cost)
+{
+	cPathCell * cell =  GetCell(a_Location);
+	if (!cell->m_IsSolid && GetCell(a_Location+Vector3d(0, -1, 0))->m_IsSolid && !GetCell(a_Location+Vector3d(0, 1, 0))->m_IsSolid)
+	{
+		ProcessCell(cell, a_Parent, a_Cost);
+	}
+}
+
+bool cPath::Step_Internal()
+{
+	cPathCell * CurrentCell = OpenListPop();
+	
+	// Path not reachable, open list exauhsted.
+	if (CurrentCell == NULL)
+	{
+		// printf("cPath::Step_Internal() - Open list is empty. Path not found.\n");
+		FinishCalculation(ePathFinderStatus::PATH_NOT_FOUND);
+		ASSERT(m_Status == ePathFinderStatus::PATH_NOT_FOUND);
+		return true;
+	}
+	
+	// Path found.
+	if (CurrentCell->m_Location == m_Destination)
+	{
+		// printf("cPath::Step_Internal() - Destination in closed list. Path Found.\n");
+		do
+		{
+			AddPoint(CurrentCell->m_Location+Vector3d(0.5, 0, 0.5));  // Populate the cPath with points.
+			CurrentCell = CurrentCell->m_Parent;
+		}
+		while (CurrentCell != NULL);
```

Comment:
Merge this line with the previous one, otherwise it's really misleading (I first thought this was an endless loop)